pkg/server/apiextensions.go

```diff
@@ -31,137 +31,232 @@ import (
+	crdLister         apiextensionslisters.CustomResourceDefinitionLister
+	workspaceLister   tenancylisters.ClusterWorkspaceLister
+	apiBindingLister  apislisters.APIBindingLister
+	systemCRDProvider *systemCRDProvider
+}
+
+var _ apiextensionslisters.CustomResourceDefinitionLister = (*apiBindingAwareCRDLister)(nil)
 
 // List lists all CustomResourceDefinitions in the underlying store matching selector. This method does not
-// support scoping to logical clusters or workspace inheritance.
-func (c *inheritanceCRDLister) List(selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+// support scoping to logical clusters or APIBindings.
+func (c *apiBindingAwareCRDLister) List(selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
 	return c.crdLister.ListWithContext(context.Background(), selector)
 }
 
-// ListWithContext lists all CustomResourceDefinitions in the logical cluster associated with ctx that match
-// selector. ClusterWorkspace API inheritance is also supported: if the ClusterWorkspace for ctx's logical cluster
-// has spec.inheritFrom set, it will aggregate all CustomResourceDefinitions from the ClusterWorkspace named
-// spec.inheritFrom with the CustomResourceDefinitions from the ClusterWorkspace for ctx's logical cluster.
-func (c *inheritanceCRDLister) ListWithContext(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+// ListWithContext lists all CustomResourceDefinitions that come in via APIBindings as well as all in the current
+// logical cluster.
+func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
 	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	// Check for API inheritance
-	inheriting := false
-	inheritFrom := ""
-	orgName, ws, err := helper.ParseLogicalClusterName(cluster.Name)
+	var ret []*apiextensionsv1.CustomResourceDefinition
+	crdName := func(crd *apiextensionsv1.CustomResourceDefinition) string {
+		return crd.Spec.Names.Plural + "." + crd.Spec.Group
+	}
+
+	// Seen keeps track of which CRDs have already been found from system and apibindings.
+	seen := sets.NewString()
+
+	org, workspace, err := helper.ParseLogicalClusterName(cluster.Name)
+	if err != nil {
+		return nil, fmt.Errorf("error determining workspace name from cluster name %q: %w", cluster.Name, err)
+	}
+
+	kcpSystemCRDs, err := c.systemCRDProvider.List(ctx, org, workspace)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving kcp system CRDs: %w", err)
+	}
+
+	// Priority 1: add system CRDs. These take priority over CRDs from APIBindings and CRDs from the local workspace.
+	ret = kcpSystemCRDs
+	for i := range kcpSystemCRDs {
+		seen.Insert(crdName(kcpSystemCRDs[i]))
+	}
+
+	apiBindings, err := c.apiBindingLister.ListWithContext(ctx, labels.Everything())
```

Comment:
(No action required) Maybe kcp listers could filter on annotations so that the switch to storing cluster name there would allow for direct filtering by that attribute?

Reply:
added todo